ipme does not detect multiple network addresses on the same interface

Test setup: have multiple ip addresses configured on the same interface:

```
$ ip addr list | grep -e 'inet ' -e ^[0-9]
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    inet 127.0.0.1/8 scope host lo
2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc pfifo_fast state DOWN group default qlen 1000
3: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    inet 172.16.198.3/16 brd 172.16.255.255 scope global noprefixroute dynamic wlan0
    inet 192.168.18.19/24 scope global wlan0
```

Ask ipmeprint about this:

```
$ ./ipmeprint 
0.0.0.0
127.0.0.1
172.16.198.3
172.16.198.3
```

Found here: https://bugs.gentoo.org/566826

I'm working on a patch.

A few amd64 results:

## macOS 10.14

```
$ sudo ifconfig lo0 inet 23.45.67.89 alias
$ sudo ifconfig en0 inet 12.34.56.78 alias

$ ifconfig -a | grep -e 'inet ' -e ^[0-9]
	inet 127.0.0.1 netmask 0xff000000
	inet 23.45.67.89 netmask 0xff000000
	inet 192.168.1.153 netmask 0xffffff00 broadcast 192.168.1.255
	inet 12.34.56.78 netmask 0xff000000 broadcast 12.255.255.255

$ git co master && make ipmeprint && ./ipmeprint
0.0.0.0
127.0.0.1
23.45.67.89
192.168.1.153
12.34.56.78

$ git co master && make ipmeprint && ./ipmeprint
0.0.0.0
127.0.0.1
23.45.67.89
192.168.1.153
12.34.56.78
```

Can't reproduce the bug. Code path: `sa_len`. If I `#undef HASSALEN` to get the `#if 1` output, it gives `0.0.0.0` and that's it.

## CentOS 6

```
$ sudo ip a add 23.45.67.89 dev lo
$ sudo ip a add 12.34.56.78 dev eth0

$ ip addr list | grep -e 'inet ' -e ^[0-9]
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN
    inet 127.0.0.1/8 scope host lo
    inet 23.45.67.89/32 scope global lo
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP qlen 1000
    inet 10.0.2.15/24 brd 10.0.2.255 scope global eth0
    inet 12.34.56.78/32 scope global eth0

$ git co master && make ipmeprint && ./ipmeprint
0.0.0.0
127.0.0.1
127.0.0.1
10.0.2.15
10.0.2.15

$ git co Dakon-ipme && make ipmeprint && ./ipmeprint
ipme.c:56: error: ‘NULL’ undeclared (first use in this function)

(change it to 0, try again)

0.0.0.0
127.0.0.1
23.45.67.89
10.0.2.15
12.34.56.78
```

Code path: non-`sa_len`. `#if 1` output looks correct.

## CentOS 7

```
$ sudo ip a add 23.45.67.89 dev lo
$ sudo ip a add 12.34.56.78 dev enp0s3

$ ip addr list | grep -e 'inet ' -e ^[0-9]
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    inet 127.0.0.1/8 scope host lo
    inet 23.45.67.89/32 scope global lo
2: enp0s3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    inet 10.0.2.15/24 brd 10.0.2.255 scope global noprefixroute dynamic enp0s3
    inet 12.34.56.78/32 scope global enp0s3

$ git co master && make ipmeprint && ./ipmeprint
0.0.0.0
127.0.0.1
127.0.0.1
10.0.2.15
10.0.2.15
```

```
$ git co Dakon-ipme && make ipmeprint && ./ipmeprint
0.0.0.0
127.0.0.1
23.45.67.89
10.0.2.15
12.34.56.78
```

Code path: non-`sa_len`. `#if 1` output looks correct.

--- a/CHANGES
+++ b/CHANGES
@@ -1,3 +1,5 @@
+20190919 fix ipme with multiple addresses on the same interface on platforms
+         without sa_len
 20190904 code: use stdint.h to get the definition for uint32
 20190903 code: do not parse header recipients for "qmail-inject -a"
 20190826 code: if $QMAILREMOTE is set, it's invoked instead of qmail-remote.

--- a/ipme.c
+++ b/ipme.c
@@ -89,6 +89,9 @@ int ipme_init()
     len = sizeof(ifr->ifr_name) + ifr->ifr_addr.sa_len;
     if (len < sizeof(*ifr))
       len = sizeof(*ifr);
+#else
+    len = sizeof(*ifr);
+#endif
     if (ifr->ifr_addr.sa_family == AF_INET) {
       sin = (struct sockaddr_in *) &ifr->ifr_addr;
       byte_copy(&ix.ip,4,&sin->sin_addr);
@@ -96,17 +99,6 @@ int ipme_init()
         if (ifr->ifr_flags & IFF_UP)
           if (!ipalloc_append(&ipme,&ix)) { close(s); return 0; }
     }
-#else
-    len = sizeof(*ifr);
-    if (ioctl(s,SIOCGIFFLAGS,x) == 0)
-      if (ifr->ifr_flags & IFF_UP)
-        if (ioctl(s,SIOCGIFADDR,x) == 0)
-	  if (ifr->ifr_addr.sa_family == AF_INET) {
-	    sin = (struct sockaddr_in *) &ifr->ifr_addr;
-	    byte_copy(&ix.ip,4,&sin->sin_addr);
-	    if (!ipalloc_append(&ipme,&ix)) { close(s); return 0; }
-	  }
-#endif
     x += len;
   }
   close(s);

--- a/ipme.c
+++ b/ipme.c
@@ -52,20 +52,35 @@ int ipme_init()
   byte_copy(&ix.ip,4,"\0\0\0\0");
   if (!ipalloc_append(&ipme,&ix)) { return 0; }
   if ((s = socket(AF_INET,SOCK_STREAM,0)) == -1) return -1;
- 
-  len = 256;
-  for (;;) {
-    if (!stralloc_ready(&buf,len)) { close(s); return 0; }
-    buf.len = 0;
+
+  ifc.ifc_buf = 0;
+  ifc.ifc_len = 0;
+
+  /* first pass: just ask what the correct length for all addresses is */
+  len = 0;
+  if (ioctl(s,SIOCGIFCONF,&ifc) >= 0 && ifc.ifc_len > 0) { /* > is for System V */
+    if (!stralloc_ready(&buf,ifc.ifc_len)) { close(s); return 0; }
     ifc.ifc_buf = buf.s;
-    ifc.ifc_len = len;
-    if (ioctl(s,SIOCGIFCONF,&ifc) >= 0) /* > is for System V */
-      if (ifc.ifc_len + sizeof(*ifr) + 64 < len) { /* what a stupid interface */
-        buf.len = ifc.ifc_len;
-        break;
-      }
-    if (len > 200000) { close(s); return -1; }
-    len += 100 + (len >> 2);
+    if (ioctl(s,SIOCGIFCONF,&ifc) >= 0)
+      buf.len = ifc.ifc_len;
+  }
+
+  /* check if we have complete length, otherwise try so sort that out */
+  if (buf.len == 0) {
+    len = 256;
+    for (;;) {
+      if (!stralloc_ready(&buf,len)) { close(s); return 0; }
+      buf.len = 0;
+      ifc.ifc_buf = buf.s;
+      ifc.ifc_len = len;
+      if (ioctl(s,SIOCGIFCONF,&ifc) >= 0) /* > is for System V */
+        if (ifc.ifc_len + sizeof(*ifr) + 64 < len) { /* what a stupid interface */
+          buf.len = ifc.ifc_len;
+          break;
+        }
+      if (len > 200000) { close(s); return -1; }
+      len += 100 + (len >> 2);
+    }
   }
   x = buf.s;
   while (x < buf.s + buf.len) {

